[WIP] Parallel builds implementation

mndeveci opened the pull request:
*Issue #, if available:*

*Why is this change necessary?*

*How does it address the issue?*

*What side effects does this change have?*

*Did you change a dependency in `requirements/base.txt`?*
    *If so, did you run `make update-reproducible-reqs`*

*Checklist:*

- [ ] Write Design Document ([Do I need to write a design document?](https://github.com/awslabs/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.rst#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [ ] `make pr` passes
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

mndeveci committed:
feat: parallel builds (utils implementation)

mndeveci committed:
feat: parallel builds (utils implementation)

mndeveci committed:
feat: parallel builds (utils implementation)

aahung committed:
Upgrade pylint (#2304)

* chore: Upgrade pylint to 2.6.0

* chore: Fix lint issues in new pylint rules

Disable import-outside-toplevel

* Add six to dependency file base.txt

mndeveci committed:
fix: use name instead of functionname when creating build graph (#2297)

mndeveci committed:
fix: for de-dup builds, build into one of the function folder instead of temporary folder (#2309)

* fix: for de-dup builds, build into one of the function folder instead of temporary folder

* fix: for de-dup builds, build into one of the function folder instead of temporary folder (fix broken unit tests in linux)

* fix: for de-dup builds, build into one of the function folder instead of temporary folder (add integration test for the usecase where Gemfile will be in the root folder)

* fix: for de-dup builds, build into one of the function folder instead of temporary folder (add integration test for the usecase where Gemfile will be in the root folder)

aahung committed:
chore: Manage black version in dev.txt (#2314)

* Move black install to dev.txt and update doc/appveyor

* Update pre-commit

* Format using black 20.8b1

* Add workaround for IDE black setup

aahung committed:
chore: Update DEVELOPMENT_GUIDE.md (typo) (#2319)

awsed committed:
Create Event for AppSync (Direct Lambda Resolver) (#2200)

* Create AppSyncDirectResolver.json

* Update AppSyncDirectResolver.json

* Update event-mapping.json

* Update event-mapping.json

Co-authored-by: Alex Wood <awood45@gmail.com>

mndeveci committed:
feat: parallel builds

mndeveci committed:
feat: parallel builds

mndeveci committed:
feat: parallel builds

mndeveci committed:
feat: parallel builds

mndeveci committed:
feat: parallel builds

mndeveci committed:
merge with cached builds

mndeveci committed:
merge with cached builds

mndeveci committed:
Merge branch 'cache-build' into parallel_builds_mingkun

mndeveci committed:
Merge branch 'cache-build' into parallel_builds_mingkun

mndeveci committed:
minor: update parallel text

mndeveci committed:
minor: update parallel text

mndeveci committed:
minor: update parallel text

hoffa commented on samcli/commands/build/command.py:
```diff
@@ -92,6 +92,12 @@
+    "--parallel",
+    "-p",
+    is_flag=True,
+    help="Use this flag to run builds of each function/layer in parallel using asyncio",
```
Is "using asyncio" relevant for users?

mndeveci replied:
Not really, it might be helpful for some customers if they want to know what is under the hood. I will check with Mark about  any change related to this

hoffa replied:
I see, it's a UX nit but I think in that case it might add cognitive overhead to users, as those unfamiliar might ask themselves "asyncio? what is it? why do I need to know about this? surely if it was mentioned I need to know about it?" etc.

mndeveci replied:
Yeah I understand your point, will raise it during doc discussions

awood45 commented on samcli/lib/build/app_builder.py:
```diff
@@ -125,9 +126,19 @@ def build(self):
         build_graph = self._get_build_graph()
-        result = self._build_functions(build_graph)
-        result.update(self._build_layers(build_graph))
+        result = self._build_functions(async_context, build_graph)
```
The way this code flow works is pretty confusing - if I'm understanding it right, this returns an empty hash when you're doing async, and in that case it updates the async_context as a side effect. It really feels like we should have a true fork in the logic with different functions in this case, this is going to be difficult to understand/test/maintain as-is in my view. At minimum we should be clearly commenting and documenting this change in approach.

awood45 commented on samcli/lib/build/app_builder.py:
```diff
@@ -242,7 +256,7 @@ def _build_unique_definition(self, build_definition):
         return function_results
 
-    def _build_layers(self, build_graph):
+    def _build_layers(self, async_context, build_graph):
```
Same comment about the significant logical fork between async and non-async.

awood45 reviewed (changes requested):
This generally looks good - however, I'm concerned with how we fork the async/non-async logic, in that the function contract changes significantly. I'd prefer a more explicit logical fork for that, even if there's a bit more verbosity.

hoffa commented on samcli/lib/utils/async_utils.py:
```diff
@@ -0,0 +1,131 @@
+    This class is used to define a task which consists of a function definition and the parameters
+    """
+
+    def __init__(self, function, args=tuple()):
+        self.function = function
+        self.args = args
```
We can use [`functool.partial()`](https://docs.python.org/3.9/library/functools.html#functools.partial) instead, e.g.

```python
task = partial(function, args)
task()
```

Allows getting rid of this class, plays well with `asyncio` (e.g. for passing keyword arguments, which isn't possible without `partial`) and simplifies `AsyncContext` (can directly append the partial function).

mndeveci committed:
- Use partial instead of custom Task class
- Update to use build strategy for Parallel builds

mndeveci committed:
- Use partial instead of custom Task class
- Update to use build strategy for Parallel builds

pyk committed:
Support running HttpApi locally (#1942)

* Add HttpApi to SamApiProvider

* Extract AWS::Serverless::HttpApi as AWS::Serverless::Api

* Implement _extract_from_serverless_http function

* Implement ApiGatewayV2LambdaEvent

* Add test for API Gateway V2 service

* Fix formatting

* Fix dedupe_function_routes

* Fix empty byte body as false

* Remove unused variables

* Remove unused variables

* Add CORS for HttpApi

* Try to increase the test timeout

* Set the timeout back to 300

* trigger build

* support for payloadFormatVersion

* support HttpApi gateway in CFN templates

* support HttpApi gateway in CFN templates - unit testing

* support HttpApi gateway in CFN templates - integration testing

* fix: 'ApiId' for httpApi is not recognized as an API in template

* support HttpApi gateway in CFN templates

* Integration tests for HttpApi support - SAM templates

* Returns json with sorted keys

* Fix failed tests

* Add 'version' field to TestApiGatewayLambdaEvent

* Black reformatting

Co-authored-by: CL-Yukky <yukihiro.yoshikawa@consallink.co.jp>
Co-authored-by: Mohamed Elasmar <melasmar@amazon.com>
Co-authored-by: Raymond Wang <14915548+wchengru@users.noreply.github.com>
Co-authored-by: Mohamed Elasmar <71043312+moelasmar@users.noreply.github.com>

wchengru committed:
apply PayloadFormationVersion in httpapi event (#2326)

mndeveci committed:
fix unit tests

moelasmar committed:
Add integration test case to test Http Api && Rest Api in the same template (#2327)

mndeveci committed:
update unit tests

mndeveci committed:
Merge branch 'cache-build' into parallel_builds_mingkun

mingkun2020 committed:
fix unit tests and add unit test for CacheBuildStrategy

mingkun2020 committed:
Merge branch 'cache-build' into mingkun_build_strategy

mingkun2020 committed:
delete unwanted changes

mingkun2020 committed:
Merge pull request #3 from mndeveci/mingkun_build_strategy

Mingkun build strategy

mingkun2020 committed:
fix CacheBuildStrategyTest after merge

mingkun2020 committed:
black reformat

mndeveci committed:
update unit tests

mndeveci committed:
update unit tests

mingkun2020 committed:
refactoring for CachedBuildStrategy unit test

mndeveci committed:
merge changes from cached build

mndeveci committed:
merge changes from cached build

mingkun2020 committed:
Merge branch 'develop' into cache-build

mndeveci committed:
Merge remote-tracking branch 'origin/parallel_builds_mingkun' into parallel_builds_mingkun

# Conflicts:
#	samcli/commands/build/command.py
#	samcli/lib/utils/async_utils.py
#	tests/integration/buildcmd/test_build_cmd.py
#	tests/unit/lib/build_module/test_app_builder.py
#	tests/unit/lib/build_module/test_build_graph.py
#	tests/unit/lib/utils/test_async_utils.py

mingkun2020 committed:
fix integration test

Commit:
fix: fail on invalid config (#2316)

mndeveci committed:
merge changes from cached build

mndeveci committed:
merge changes from cached build

hoffa commented:
LGTM 👌

mingkun2020 committed:
fix integration test import

mingkun2020 committed:
Merge branch 'develop' into cache-build

mingkun2020 committed:
fix unittest in test_app_builder

mingkun2020 committed:
Merge branch 'cache-build' of https://github.com/mingkun2020/aws-sam-cli into cache-build

mingkun2020 committed:
update constructor of build graph to python 3 style and black reformat

CoshUS committed:
Fix: Missing Unix Permission Bits for Artifact ZIP Created in Windows (#2193)

* Added Mock Unix Permission for ZIP Created in Windows

* Fixed Import Order

* Changed Permission from 755 to 005

* Added make_zip Windows Unit Test

* Updated Comment

mingkun2020 committed:
update code based on review

mingkun2020 committed:
update unit test

CoshUS committed:
Fix: relpath Fails with Different Drives (#2138)

* Fixed relpath Fails with Different Drives

* Fixed pylint Formatting

* Added Unit Test for AppBuilder with Different Drives

* Changed absolute path unittest to Windows Specific

* Reformatted

* Removed Unused Import

mingkun2020 committed:
update unit test for build_strategy

mndeveci committed:
Merge branch 'develop' into cache-build

mingkun2020 committed:
update unit test

svenemtell committed:
Fix: Make API Gateway infer Lambda function response for HttpApi with PayloadFormatVersion 2.0 (#2334)

* Make API Gateway infer Lambda function response for HttpApi with PayloadFormatVersion 2.0.

* Remove unnecessary empty line.

* check the payload format version before handling it in version 2 format

* lambda response can be any valid json in payload format v2.0

* add integration test cases for lambda response payload format v2.0

* fix lint error

* add more unit test cases

* fix the default route handling

Co-authored-by: Mohamed Elasmar <71043312+moelasmar@users.noreply.github.com>
Co-authored-by: Mohamed Elasmar <melasmar@amazon.com>

mndeveci committed:
merge changes with cache build

wchengru committed:
chore: bump sam cli version to 1.7.0, and sam translator requirement to 1.28.1 (#2336)

Co-authored-by: Wing Fung Lau <4760060+hawflau@users.noreply.github.com>

mndeveci committed:
Merge remote-tracking branch 'origin/parallel_builds_mingkun' into parallel_builds_mingkun

mndeveci committed:
Merge branch 'develop' into parallel_builds_mingkun

mndeveci committed:
black changes

mingkun2020 committed:
feat: Cache build (#2295)

* chore: bump lambda-builder version v1.1.0 and sam-cli version v1.2.0.

* after updating version, run a command to update requirements/reproducible-linux.txt file

* De-duplicated builds

* De-duplicated builds (updated unit tests)

* De-duplicated builds (add actual build functionality and related unit tests)

* De-duplicated builds (build graph should be created inside build folder)

* De-duplicated builds (update tests with os specific separator)

* De-duplicated builds (update tests with os specific separator)

* De-duplicated builds (update tests with os specific separator)

* De-duplicated builds (update tests with os specific separator)

* De-duplicated builds (added integration tests)

* De-duplicated builds

- Changed to add functions to use dedicated method
- Added extra checks when function name or handler name is called
- Removed creation of tmp directory from one of the unit test
- Added validation for dedup integ tests to validate functions build only once

* De-duplicated builds

- Changed to identify each build as unique for custom Makefile builds
- Added unit tests related to this change
- Added integration tests related to this change

* De-duplicated builds

- Update integration tests to run each couple of functions because of timeout

* De-duplicated builds

- Update is_building_specific_resource function name and add documentation for it
- Updated formatting of some files

* De-duplicated builds

- Update documentation and match parameter names for specific build function

* implement cache build

* change attribute name from checksum to source_md5

* update logic for rm cache directory

* update unit test for build graph

* refactor build definition and update unit test

* update unit test

* bug fix for the cached flag property

* add integration test for cache build

* Update documentation and log message

* add cached build for layers

* update comment for layer cached build

* small refactoring

* feat: build strategy

* feat: build strategy

* fix unit tests

* update unit tests

* fix unit tests and add unit test for CacheBuildStrategy

* delete unwanted changes

* fix CacheBuildStrategyTest after merge

* black reformat

* refactoring for CachedBuildStrategy unit test

* fix integration test

* fix integration test import

* fix unittest in test_app_builder

* update constructor of build graph to python 3 style and black reformat

* update code based on review

* update unit test

* update unit test for build_strategy

* update unit test

* update docs and replace getter/setter with direct attr access

* update comments

Co-authored-by: Mehmet Nuri Deveci <5735811+mndeveci@users.noreply.github.com>

mndeveci committed:
Merge branch 'develop' into parallel_builds_mingkun

mingkun2020 commented:
This issue has not received any attention in 1 year. If you want to keep this issue open, please leave a comment below and auto-close will be canceled.
